File: src/todo/domain/ports.py
"""Ports: the interfaces the application depends on but does not implement.

These exist so the service can be driven at speed in tests (in-memory repository,
frozen clock) and in earnest in production (SQLite, the real clock) without the
service knowing which it got.
"""

from __future__ import annotations

from datetime import datetime, timezone
from typing import Protocol

from .model import Todo


class TodoRepository(Protocol):
    def add(self, todo: Todo) -> None: ...

    def replace(self, todo: Todo) -> None: ...

    def remove(self, todo_id: str) -> None: ...

    def get(self, todo_id: str) -> Todo | None: ...

    def find_active_by_title(self, title: str) -> Todo | None: ...

    def all(self) -> list[Todo]: ...


class Clock(Protocol):
    def now(self) -> datetime: ...


class SystemClock:
    def now(self) -> datetime:
        return datetime.now(timezone.utc)


class FrozenClock:
    """A clock the tests control.

    Time is an input to the system like any other. Injecting it is what makes
    'a todo completed yesterday sorts below one completed today' a testable
    statement instead of a sleep().
    """

    def __init__(self, start: datetime) -> None:
        self._now = start

    def now(self) -> datetime:
        return self._now

    def advance(self, **delta) -> None:
        from datetime import timedelta

        self._now = self._now + timedelta(**delta)

    def set(self, moment: datetime) -> None:
        self._now = moment

File: haskell/src/Todo/Domain/Ports.hs
-- | Ports: the interfaces the application depends on but does not implement.
--
-- Records of 'IO' actions are the Haskell equivalent of Python's @Protocol@ —
-- the service is handed one and never learns whether it is talking to an
-- in-memory map or SQLite, a frozen clock or the wall clock.
module Todo.Domain.Ports
  ( TodoRepository (..)
  , Clock (..)
  , systemClock
  , FrozenClock (..)
  , newFrozenClock
  ) where

import Data.IORef (modifyIORef', newIORef, readIORef, writeIORef)
import Data.Text (Text)
import Data.Time (NominalDiffTime, UTCTime, addUTCTime, getCurrentTime)

import Todo.Domain.Model (Todo)

data TodoRepository = TodoRepository
  { repoAdd               :: Todo -> IO ()
  , repoReplace           :: Todo -> IO ()
  , repoRemove            :: Text -> IO ()
  , repoGet               :: Text -> IO (Maybe Todo)
  , repoFindActiveByTitle :: Text -> IO (Maybe Todo)
  , repoAll               :: IO [Todo]
  }

newtype Clock = Clock { clockNow :: IO UTCTime }

systemClock :: Clock
systemClock = Clock getCurrentTime

-- | A clock the tests control. Time is an input like any other; injecting it is
-- what makes "finished yesterday sorts below finished today" a testable
-- statement instead of a @sleep@.
data FrozenClock = FrozenClock
  { frozenClock  :: Clock
  , advanceClock :: NominalDiffTime -> IO ()
  , setClock     :: UTCTime -> IO ()
  }

newFrozenClock :: UTCTime -> IO FrozenClock
newFrozenClock start = do
  ref <- newIORef start
  pure FrozenClock
    { frozenClock  = Clock (readIORef ref)
    , advanceClock = \delta -> modifyIORef' ref (addUTCTime delta)
    , setClock     = writeIORef ref
    }
